### ISSUE TITLE
toDevice messages on matrix.org can take 14.1s to propagate.

### ISSUE THREAD

**Issue by ara4n** [matrix-org/synapse#9564]
https://github.com/vector-im/element-web/issues/8376

This has been around for years, but bites me surprisingly frequently (even on test accounts!) making me assume it might be quite a widespread problem (and perhaps contribute to UISIs and encryption delays).

**Comment by anoadragon453** [comment]
Note that the linked Element Web issue has Synapse logs as well. At the moment it looks to have been tracked down to the toDevice message not waking up the synchrotron for ~14s. We do at least *try* to do so:

https://github.com/matrix-org/synapse/blob/fc8b3d88097d2c985e8eae9779cdb4b23e1a8ef6/synapse/handlers/devicemessage.py#L237-L239

We also talked about increasing Jaeger's monitoring on matrix.org to try and catch the full path of the request as it goes through the system (it's currently only at 10%). We also don't have the encryption worker in Jaeger's services at the moment, which would be a good initial fix to begin with.

**Comment by callahad** [comment]
From the team sync today, we think getting our Jaeger deployment into a place where it's useful for solving this could turn into a pretty significant rabbit hole, and we're at capacity for rabbit holes at the moment.

The hope is that we'll be able to come up for air in the next ~1-2 weeks and consider digging in more deeply, but in the meantime ara4n is going to do some manual digging and see if he turns up anything useful.

**Comment by richvdh** [comment]
related: #9533

**Comment by richvdh** [comment]
@ara4n reported another delayed message:
```
Wed May 12 16:43:29 BST 2021
1620834209.501334000
{}
->

16:43:35.609 rageshake.js:65 received to_device net.arasphere.test from: @matthewtest7:matrix.org id: undefined
```

On the encryption worker:
```
2021-05-12 15:43:29,579 - synapse.9533_debug - 549 - DEBUG - PUT-316121-- - Stored to-device messages with stream_id 260733698 for [('@matthewtest7:matrix.org', '29yr+WWkOOlgiOJH+aVIzlejMCeGBkUBLSH4VolH2o8'), ('@matthewtest7:matrix.org', 'AGMPOJNUXK'), ('@matthewtest7:matrix.org', 'BvrEPEYN5xqarqzvw15TawHRGopBlCZB4BNt+wTcX20'), ('@matthewtest7:matrix.org', 'DHKBGEETBN'), ('@matthewtest7:matrix.org', 'FBR5CDAZUUmdxU0w6kkLkARm+OGN5mOQq46D/fHHNfE'), ('@matthewtest7:matrix.org', 'FIGEQEAQRK'), ('@matthewtest7:matrix.org', 'FVRJKEZQBW'), ('@matthewtest7:matrix.org', 'GAWSPZDHZY'), ('@matthewtest7:matrix.org', 'GCEZUQZUQO'), ('@matthewtest7:matrix.org', 'GEFBWQHMEO'), ('@matthewtest7:matrix.org', 'HNGIOKUKWZ'), ('@matthewtest7:matrix.org', 'IETEMEMWSB'), ('@matthewtest7:matrix.org', 'KAYDNSRDZZ'), ('@matthewtest7:matrix.org', 'LGFPVHDGHM'), ('@matthewtest7:matrix.org', 'MROFBYOBJN'), ('@matthewtest7:matrix.org', 'NBi2sj9rh/xflRIjIdUdp8iRpOSHjeiXyPHoWofDcnw'), ('@matthewtest7:matrix.org', 'PTSANZNJNZ'), ('@matthewtest7:matrix.org', 'QDTBLOQDCO'), ('@matthewtest7:matrix.org', 'RJEEKTTNAD'), ('@matthewtest7:matrix.org', 'RSACOJTTMA'), ('@matthewtest7:matrix.org', 'TBOTWOIPUO'), ('@matthewtest7:matrix.org', 'TMEBWPPFVP'), ('@matthewtest7:matrix.org', 'UlPr+z2c2tzBAVO1NJED+Cdz9Ib86Oax2c2uDCLqtdA'), ('@matthewtest7:matrix.org', 'VEVLFOCPIL'), ('@matthewtest7:matrix.org', 'WN4TmF8uxs7uqdOCv8RxqQFqBzg4c9UaelAhd4cgXTU'), ('@matthewtest7:matrix.org', 'WQNXLQTROL'), ('@matthewtest7:matrix.org', 'YKFVSQHZTO'), ('@matthewtest7:matrix.org', 'YXDAYDVKRF'), ('@matthewtest7:matrix.org', 'Z6Ws6GEVaHzgN5lC3aoK6MAwPXY/sHRF/F74fw8X8Yk'), ('@matthewtest7:matrix.org', 'cesMHMyNR5kLVE/4+18pfbdn+arMLmLXds9heiPn/jw'), ('@matthewtest7:matrix.org', 'gmNxf99ZtaMc11G66UM8J3bhDMBk93GFtAhGgfNEHTI'), ('@matthewtest7:matrix.org', 'iW2mIVnEkmy/P6TT6Cyq9s42JBfAFSARUQzLp3TIG7s'), ('@matthewtest7:matrix.org', 'j/BtV3WT8au+CKcsYXgLPYtKf4cLUgFyLkzv2qffOiE'), ('@matthewtest7:matrix.org', 'mRLJfUT2SoChNJ8ZLqMOwiAwjR4yKzSKoe6AM4U6l9A')]
2021-05-12 15:43:29,582 - synapse.9533_debug - 434 - DEBUG - PUT-316121- - to-device messages stream id 260733698, awaking streams for dict_keys(['@matthewtest7:matrix.org'])
2021-05-12 15:43:29,583 - synapse.access.http.8134 - 408 - INFO - PUT-316121 - 80.250.101.234 - 8134 - {@matthewtest7:matrix.org} Processed request: 0.014sec/-0.000sec (0.003sec, 0.001sec) (0.001sec/0.010sec/3) 2B 200 "PUT /_matrix/client/r0/sendToDevice/net.arasphere.test/1620834209.501334000 HTTP/1.1" "curl/7.64.1" [0 dbevts]
```

On synchrotron8:

```
2021-05-12 15:43:29,610 - synapse.9533_debug - 434 - DEBUG - process-replication-data-2994089 - to-device messages stream id 260733698, awaking streams for ['@matthewtest7:matrix.org']
2021-05-12 15:43:32,845 - synapse.9533_debug - 434 - DEBUG - process-replication-data-2994483 - to-device messages stream id 260733716, awaking streams for ['@matthewtest7:matrix.org']
2021-05-12 15:43:32,875 - synapse.9533_debug - 434 - DEBUG - process-replication-data-2994530 - to-device messages stream id 260733718, awaking streams for ['@matthewtest7:matrix.org']
2021-05-12 15:43:32,896 - synapse.9533_debug - 434 - DEBUG - process-replication-data-2994577 - to-device messages stream id 260733719, awaking streams for ['@matthewtest7:matrix.org']
2021-05-12 15:43:35,123 - synapse.access.http.8300 - 408 - INFO - GET-581195 - 80.250.101.234 - 8300 - {@matthewtest7:matrix.org} Processed request: 6.068sec/-0.000sec (0.117sec, 0.002sec) (0.986sec/0.479sec/7) 1458B 200 "GET /_matrix/client/r0/sync?filter=4&timeout=30000&since=s2009948967_757284961_681796_864628704_759364233_2306590_260733689_1011208698_195525 HTTP/1.1" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.93 Safari/537.36" [1 dbevts]
```

so we can tell that the replication notification is making it as far as the synchrotron, but the /sync request isn't waking up for some reason (or it took 6 seconds between wakeup and return). Better add some more debug...

**Comment by richvdh** [comment]
I think this is basically fixed by #10124, though we'll leave some of the debugging in place to help trace it if it turns up again.
